kms/cloudkms/cloudkms.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	var signatureAlgorithm kmspb.CryptoKeyVersion_CryptoKeyVersionAlgorithm
+	v, ok := signatureAlgorithmMapping[req.SignatureAlgorithm]
+	if !ok {
+		return nil, errors.Errorf("cloudKMS does not support signature algorithm '%s'", req.SignatureAlgorithm)
+	}
+	switch v := v.(type) {
+	case kmspb.CryptoKeyVersion_CryptoKeyVersionAlgorithm:
+		signatureAlgorithm = v
+	case map[int]kmspb.CryptoKeyVersion_CryptoKeyVersionAlgorithm:
+		if signatureAlgorithm, ok = v[req.Bits]; !ok {
+			return nil, errors.Errorf("cloudKMS does not support signature algorithm '%s' with '%d' bits", req.SignatureAlgorithm, req.Bits)
+		}
+	default:
+		return nil, errors.Errorf("unexpected error: this should not happen")
+	}
+
+	var crytoKeyName string
+
+	// Split `projects/PROJECT_ID/locations/global/keyRings/RING_ID/cryptoKeys/KEY_ID`
+	// to `projects/PROJECT_ID/locations/global/keyRings/RING_ID` and `KEY_ID`.
+	keyRing, keyID := Parent(req.Name)
+	if err := k.createKeyRingIfNeeded(keyRing); err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := defaultContext()
+	defer cancel()
+
+	// Create private key in CloudKMS.
+	response, err := k.client.CreateCryptoKey(ctx, &kmspb.CreateCryptoKeyRequest{
+		Parent:      keyRing,
+		CryptoKeyId: keyID,
+		CryptoKey: &kmspb.CryptoKey{
+			Purpose: kmspb.CryptoKey_ASYMMETRIC_SIGN,
+			VersionTemplate: &kmspb.CryptoKeyVersionTemplate{
+				ProtectionLevel: protectionLevel,
+				Algorithm:       signatureAlgorithm,
+			},
+		},
+	})
+	if err != nil {
+		if status.Code(err) != codes.AlreadyExists {
+			return nil, errors.Wrap(err, "cloudKMS CreateCryptoKey failed")
+		}
+		// Create a new version if the key already exists.
+		//
+		// Note that it will have the same purpose, protection level and
+		// algorithm than as previous one.
+		req := &kmspb.CreateCryptoKeyVersionRequest{
+			Parent: req.Name,
+			CryptoKeyVersion: &kmspb.CryptoKeyVersion{
+				State: kmspb.CryptoKeyVersion_ENABLED,
+			},
+		}
+		response, err := k.client.CreateCryptoKeyVersion(ctx, req)
+		if err != nil {
+			return nil, errors.Wrap(err, "cloudKMS CreateCryptoKeyVersion failed")
+		}
+		crytoKeyName = response.Name
+	} else {
+		crytoKeyName = response.Name + "/cryptoKeyVersions/1"
+	}
+
+	// Retrieve public key to add it to the response.
+	pk, err := k.GetPublicKey(&apiv1.GetPublicKeyRequest{
+		Name: crytoKeyName,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "cloudKMS GetPublicKey failed")
+	}
+
+	return &apiv1.CreateKeyResponse{
+		Name:      crytoKeyName,
+		PublicKey: pk,
+		CreateSignerRequest: apiv1.CreateSignerRequest{
+			SigningKey: crytoKeyName,
+		},
+	}, nil
+}
+
+func (k *CloudKMS) createKeyRingIfNeeded(name string) error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+
+	_, err := k.client.GetKeyRing(ctx, &kmspb.GetKeyRingRequest{
+		Name: name,
+	})
+	if err == nil {
+		return nil
+	}
+
+	parent, child := Parent(name)
+	_, err = k.client.CreateKeyRing(ctx, &kmspb.CreateKeyRingRequest{
```

Comment:
Why not just run CreateKeyRing (instead of GetKeyRing), since you're still checking for the AlreadyExistsError?

Reply:
Well, let's say just in case, better to be safe than sorry.